internal/core/src/segcore/InsertRecord.h

```diff
@@ -86,6 +95,70 @@ class OffsetHashMap : public OffsetMap {
     std::unordered_map<T, std::vector<int64_t>> map_;
 };
 
+template <typename T>
+class OffsetOrderedArray : public OffsetMap {
+ private:
+    int
+    find(const T pk) const {
+        int l = 0, r = array_.size() - 1;
+        while (l < r) {
+            int m = (l + r) >> 1;
+            if (array_[m].first < pk)
+                l = m + 1;
+            else
+                r = m;
+        }
+        return r;
+    }
+
+ public:
+    std::vector<SegOffset>
+    find_with_timestamp(const PkType pk, Timestamp timestamp, const ConcurrentVector<Timestamp>& timestamps) const {
+        std::vector<SegOffset> res_offsets;
```

Comment:
construct this vector with iterator after replace `find` with `std::binary_search()`